File: java/core/src/core/mail/client/Master.java
/**
 * Author: Timothy Prepscius
 * License: GPLv3 Affero + keep my name in the code!
 */
package mail.client;

import mail.client.cache.JSON;
import mail.client.model.AddressBook;
import mail.client.model.Identity;
import core.crypt.Cryptor;
import core.util.Environment;

public class Master 
{
	protected Initializer initializer;
	protected Store store;
	protected Identity identity;
	protected Environment environment;
	protected AddressBook addressBook;
	protected Indexer indexer;
	protected ArrivalsProcessor arrivalsProcessor;
	protected ArrivalsMonitor arrivalsMonitor;
	protected EventDispatcher eventDispatcher;
	protected Actions actions;
	protected Mailer mailer;
	protected Cryptor cryptor;
	protected CacheManager cacheManager;
	protected JSON json;
	
	public Master (
		Store store,
		Identity identity,
		Environment environment,
		Indexer indexer, 
		AddressBook addressBook,
		ArrivalsProcessor arrivalsProcessor, 
		ArrivalsMonitor arrivalsMonitor, 
		EventDispatcher eventDispatcher,
		Actions actions,
		Mailer mailer,
		Cryptor cryptor,
		CacheManager cacheManager,
		JSON json
	)
	{
		this.identity = identity;
		this.environment = environment;
		this.addressBook = addressBook;

		this.store = store;
		store.setMaster(this);

		this.initializer = new Initializer();
		initializer.setMaster(this);

		this.indexer = indexer;
		indexer.setMaster(this);

		this.arrivalsProcessor = arrivalsProcessor;
		arrivalsProcessor.setMaster(this);

		this.arrivalsMonitor = arrivalsMonitor;
		arrivalsMonitor.setMaster(this);

		this.eventDispatcher = eventDispatcher;

		this.actions = actions;
		actions.setMaster(this);

		this.mailer = mailer;
		mailer.setMaster(this);

		this.cryptor = cryptor;

		this.cacheManager = cacheManager;
		cacheManager.setMaster(this);

		this.json = json;
		json.setMaster(this);
	}
	
	public void start() throws Exception
	{
		initializer.start();
	}
	
	public Store getStore ()
	{
		return store;
	}
	
	public Identity getIdentity ()
	{
		return identity;
	}
	
	public Environment getEnvironment ()
	{
		return environment;
	}
	
	public AddressBook getAddressBook ()
	{
		return addressBook;
	}
	
	public Indexer getIndexer()
	{
		return indexer;
	}
	
	public CacheManager getCacheManager()
	{
		return cacheManager;
	}
	
	public ArrivalsProcessor getArrivalsProcessor()
	{
		return arrivalsProcessor;
	}

	public ArrivalsMonitor getArrivalsMonitor() 
	{
		return arrivalsMonitor;
	}

	public EventDispatcher getEventPropagator ()
	{
		return eventDispatcher;
	}
	
	public Actions getActions ()
	{
		return actions;
	}
	
	public Mailer getMailer ()
	{
		return mailer;
	}
	
	public Cryptor getCryptor ()
	{
		return cryptor;
	}
	
	public JSON getJSON ()
	{
		return json;
	}
}

File: cpp/core/mailiverse/mail/manager/Master.cpp
/**
 * Author: Timothy Prepscius
 * License: BSD + keep my name in the code!
 */

#include "Master.h"
#include "Lib.h"
#include "mailiverse/core/connector/Lib.h"
#include "mailiverse/core/store/Lib.h"
#include "mailiverse/core/crypt/Lib.h"
#include "EventDispatcherThreaded.h"
#include "AsyncActions.h"

using namespace mailiverse::mail::manager;
using namespace mailiverse::core::connector;
using namespace mailiverse::core::constants;
using namespace mailiverse::mail::model;
using namespace mailiverse::core::store;
using namespace mailiverse::core::crypt;
using namespace mailiverse::utilities;
using namespace mailiverse;

Master *Master::create (const String &identityString, core::store::Environment *environment)
{
	EnvironmentPtr clientEnvironment = environment->childEnvironment(ConstantsEnvironmentKeys::CLIENT);
	TrackingConnector *connector = new TrackingConnector(StoreConnectorFactory::create(*clientEnvironment));
	
	AddressBookPtr addressBook = new AddressBook();
	IdentityPtr identity = addressBook->getIdentity(
		new UnregisteredIdentity(identityString)
	);
	identity->setPrimary(true);
	
	CryptorRSAPtr cryptorRSA = CryptorRSAFactory::create(*clientEnvironment);
	CryptorRSAAESPtr cryptorRSAAES = new CryptorRSAAES(cryptorRSA);
	CryptorSeed cryptorSeed(Base64::decode(toString(cryptorRSA->getPrivateKey())));

	Master *master = new Master (
		new manager::Store(cryptorRSAAES, connector),
		identity,
		clientEnvironment,
		new Indexer(),
		addressBook,
		new ArrivalsProcessor(),
		new ArrivalsMonitorDefault(cryptorRSAAES, connector),
		new EventDispatcherThreaded(),
		new AsyncActions(),
		new Mailer(NULL),
		cryptorRSAAES,
		new CacheManager(cryptorSeed, connector)
	);

	connector->setMaster(master);
	
	return master;
}
